src/crush/CrushCompiler.cc

```diff
@@ -240,14 +244,28 @@ int CrushCompiler::decompile(ostream &out)
   }
 
   out << "\n# rules\n";
-  for (int i=0; i<crush.get_max_rules(); i++) {
-    if (!crush.rule_exists(i))
+  std::set<int> rules;
+  for (int ri=0; ri<crush.get_max_rules(); ri++) {
+    if (!crush.rule_exists(ri))
+      continue;
+    int i;
+    if (crush.original_rule.find(ri) != crush.original_rule.end())
+      // we found a rule that was generated from anothe rule
+      i = crush.original_rule[ri];
+    else
+      i = ri;
+    if (rules.find(i) == rules.end())
+      rules.insert(i);
+    else
       continue;
     out << "rule ";
     if (crush.get_rule_name(i))
       print_rule_name(out, i, crush);
     out << " {\n";
-    out << "\truleset " << crush.get_rule_mask_ruleset(i) << "\n";
+    // Use ri instead of i here: we need the ruleno of the generated
+    // rule because it is exposed to the user. The original rule is
+    // given another ruleno, never exposed to the user.
+    out << "\truleset " << crush.get_rule_mask_ruleset(ri) << "\n";
```

Comment:
do we actually need to generate new rules?  i think the original rule can simply use the generated item for the take call, and when decompiled, the fact that the item is generated will tell us to add the 'class xxx' argument to the take step.